Design Document: Omega-0 algorithm formulation

mark-petersen opened the pull request:
Add section 3, the algorithm formulation, to the design document. During this review, the documentation is [available here](https://portal.nersc.gov/project/e3sm/mpeterse/omega/design-omega0-sw/html/design/OmegaV0ShallowWater.html) for convenient viewing.

Checklist
* [x] Documentation:
  * [x] Documentation has been [built locally](https://e3sm-project.github.io/omega/develop/developers_guide/building_docs.html) and changes look as expected

xylar commented:
@mark-petersen, as I mentioned in #32, please keep the checklist when you make new Omega PRs.

mark-petersen commented:
This PR is still under development. Early comments are welcome, but I will add reviewers and post again when the algorithm section is complete.

xylar commented:
@mark-petersen, it doesn't look like you rebased onto the current `develop`, so you have redundant commits from #32 that I don't think we want.

hyungyukang commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,537 @@
+
+$$
+\frac{\partial \boldsymbol{u}}{\partial t} + \left(\boldsymbol{u}\cdot\nabla\right)\boldsymbol{u}  + f \boldsymbol{k}\times\boldsymbol{u}
+= -g\nabla(h+b)
```
Shouldn't `(h+b)` be `(h+H)` according to the definitions of `H` in the table below? Or need to change `H` in the table to `b` as bottom topography?

mark-petersen replied:
Done. I changed H to b.

mark-petersen commented:
During this review, the documentation is [available here](https://portal.nersc.gov/project/e3sm/mpeterse/omega/design-omega0-sw/html/design/OmegaV0ShallowWater.html) for convenient viewing.

mark-petersen commented:
Rebased and squashed. This is document is now done and ready for review.

mark-petersen commented:
Please review when you get a chance: @sbrus89 @mwarusz @hyungyukang 

I would like to make any revisions and merge this week.

sbrus89 commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,607 @@
+
+Omega mesh information will be stored in a separate mesh file, and not be included with initial condition, restart, or output files. This way there is never redundant mesh data stored in these files.
+
+To do: In the design section below, add list of minimum number of variables that defines the mesh, and the additional variables that can be computed on start-up.
```
Do we want to still do this?

mark-petersen replied:
Good point. I think that is not needed here.

sbrus89 commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,607 @@
+We begin with the standard shallow water equations, which are derived from conservation of momentum and mass in a single layer with uniform density in a rotating frame. The standard presentation may be found in standard textbooks on Geophysical Fluid Dynamics, such as those by Vallis (2017), Cushman‐Roisin and Beckers (2011), Pedlosky (1987), and Gill (2016). In continuous form, these are
+
+$$
+\frac{\partial \boldsymbol{u}}{\partial t} + \left(\boldsymbol{u}\cdot\nabla\right)\boldsymbol{u}  + f \boldsymbol{k}\times\boldsymbol{u}
+= -g\nabla(h+b)
```
Do we want $b$ to follow the traditional positive-down convention since that's what we use in the code? It's also fine to leave as-is since this is consistent with Ringler 2010.

mark-petersen replied:
Good point. I changed to `-b` so it matches the `BottomDepth` variable.

sbrus89 commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,607 @@
+| $f$                 | Coriolis parameter          | 1/s      | vertex   | fVertex  |                                                              |
+| $\mathcal{F} $      | forcing                     | m/s^2    | edge     |   |                                                              |
+| $g$                 | gravitational acceleration  | m/s^2    | constant |   |                                                              |
+| $h$                 | thickness of layer          | m        | cell     | layerThickness  |                                                              |
```
OMEGA variable names start with a capital letter

mark-petersen replied:
Thanks. I made that change. The variable names really go in the design section below, but it is so convenient to put here that I started to fill it in.

sbrus89 commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,607 @@
+$$
+\nabla \cdot \left( u_e [h_i \phi_i]_e \right)  =  \frac{1}{A_i} \sum_{e\in EC(i)} n_{e,i}u_e [h_i \phi_i]_e l_e
+$$
+The question is how to interpolate to the edge to obtain $[h_i \phi_i]_e$. Centered advection simply uses the midpoint rule. Upwind takes the value from the upstream cell. Higher order schemes cast a wider stencil to compute the value at this edge.
```
We should add the higher order scheme here down the road.

mark-petersen replied:
Agreed. I added a sentence that MPAS-Ocean used FCT.

sbrus89 reviewed (commented):
@mark-petersen, this looks great to me. I just had a couple really minor comments from reading through.

mwarusz reviewed (changes requested):
Looks good @mark-petersen. I also only have minor comments.

hyungyukang reviewed (approved):
@mark-petersen , changes look great. @sbrus89 and @mwarusz 's reviews covered my suggestions and improved the doc. Many thanks for your work!

hyungyukang commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,593 @@
+| $f$                 | Coriolis parameter          | 1/s      | vertex   |   |                                                              |
+| $\mathcal{F} $      | forcing                     | m/s^2    | edge     |   |                                                              |
+| $g$                 | gravitational acceleration  | m/s^2    | constant |   |                                                              |
+| $H$                 | total unperturbed depth     | m        | cell     |   |                                                              |
```
This suggestion is resolved by @mark-petersen .

mark-petersen commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,607 @@
+| $i\in CE(e)$ |  Two primal mesh cells that share edge $e$ | cellsOnEdge |
+| $i\in CV(v)$ |  Set of primal mesh cells that form the vertices of dual mesh cell $D_v$ | cellsOnVertex |
+| $v\in VE(e)$ |  The two dual-mesh cells that share edge $e$ | verticesOnEdge |
+| $v\in VI(i)$ |  The set of dual-mesh cells that form the vertices of primal mesh cell $P_i$ | verticesOnCell |
```
Using the notation in [Ringler 2010](https://www.sciencedirect.com/science/article/pii/S0021999109006780#tbl2), the `i` is used for cell center. I think they chose that because subscript `c` and `e` look the same.

mwarusz replied:
@mark-petersen I am not insisting on this change, but I think that this is simply a typo in the table in Ringler 2010. See Eq (26) therein where this variable is called `VC` and Table 2.2 in a later [TRiSK extension paper](https://doi.org/10.1137/110850293).

mark-petersen replied:
Thanks. You are right, I see that now. It was also correct in the original [Thuburn et al. 2009](http://www.sciencedirect.com/science/article/pii/S0021999109004434).

mark-petersen commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,607 @@
+The Del2, or Laplacian operator, viscous momentum dissipation maps edge-normal velocity back to the edge-normal component of the Laplacian. In TRiSK this is done with the "grad-div plus vorticity curl" formulation. (check grad omega term).
+
+$$
+\nu_2 \nabla^2 u_e = \nu_2 \left( \nabla \delta_i + {\bf k} \times \nabla \omega_v\right) = \frac{\nu_2}{d_e} \sum_{i\in CE(e)} -n_{e,i}\delta_i + \frac{\nu_2}{l_e} \sum_{v\in VE(e)} -n_{e,v}\omega_v
```
Thanks. That is divergence. I will add it.

sbrus89 reviewed (approved):
Thanks for addressing those changes, @mark-petersen. This is a really nice resource to have!

mark-petersen commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,607 @@
+[Ringler et al. 2010](https://www.sciencedirect.com/science/article/pii/S0021999109006780).
+
+#### 3.2.5. Perpendicular Gradient
+This maps a vertex scalar to an edge-tangential vector component (check $t_{e,v}$)
```
Thanks. I updated the document to explain this, and added the variable to the table.

mark-petersen commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,607 @@
+$$
+
+#### 3.3.4. Del2 momentum dissipation
+The Del2, or Laplacian operator, viscous momentum dissipation maps edge-normal velocity back to the edge-normal component of the Laplacian. In TRiSK this is done with the "grad-div plus vorticity curl" formulation. (check grad omega term).
```
Thanks. I updated the document to explain this

mark-petersen commented:
@mwarusz I corrected based on your comments. The [full document is here](https://portal.nersc.gov/project/e3sm/mpeterse/omega/design-omega0-sw/html/design/OmegaV0ShallowWater.html).

In particular, I expanded the [section on the perpendicular gradient](https://portal.nersc.gov/project/e3sm/mpeterse/omega/design-omega0-sw/html/design/OmegaV0ShallowWater.html#perpendicular-gradient) and the [section on del2](https://portal.nersc.gov/project/e3sm/mpeterse/omega/design-omega0-sw/html/design/OmegaV0ShallowWater.html#del2-momentum-dissipation).

Please review again if you like, and I can merge in once you've approved.

mwarusz commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,660 @@
+\equiv
+\nabla h_i = \frac{h_{i2} - h_{i1} }{d_e},
+$$
+where $\left\{i_1, i_2\right\}$ are the cells neighboring edge $e$.  The indices $\left\{i_1, i_2\right\}$ are ordered such that the normal vector ${\bf n}$ points from cell $i_1$ to cell $i_2$. In the code ${\bf n}$ points from CellsOnEdge(0,iEdge) to CellsOnEdge(1,iEdge).
```
```suggestion
where $\left\{i_1, i_2\right\}$ are the cells neighboring edge $e$.  The indices $\left\{i_1, i_2\right\}$ are ordered such that the normal vector ${\bf n}$ points from cell $i_1$ to cell $i_2$. In the code ${\bf n}$ points from CellsOnEdge(IEdge, 0) to CellsOnEdge(IEdge, 1).
```

mwarusz commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,660 @@
+$$
+\left(\nabla \omega_v\right)_e^\perp = \frac{\omega_{v2} - \omega_{v1}}{l_e},
+$$
+where the positive vector ${\bf n}^\perp$ is 90$^o$ to the left of ${\bf n}$. The indices are ordered such that ${\bf n}^\perp$ points from $v_1$ to $v_2$, which corresponds to VerticesOnEdge(0,iEdge) and VerticesOnEdge(1,iEdge) in the code.
```
```suggestion
where the positive vector ${\bf n}^\perp$ is 90$^o$ to the left of ${\bf n}$. The indices are ordered such that ${\bf n}^\perp$ points from $v_1$ to $v_2$, which corresponds to VerticesOnEdge(IEdge, 0) and VerticesOnEdge(IEdge, 1) in the code.
```

mwarusz commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,660 @@
+[Ringler et al. 2010](https://www.sciencedirect.com/science/article/pii/S0021999109006780) equation 63,
+
+$$
+K_i = \frac{1}{A_i} \sum_{e\in EC(c)} \frac{{\bar A}_{e}}{4}u_e^2
```
`A_e` is not defined and is missing from the variables table.

mark-petersen replied:
done, thanks.

mwarusz commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,660 @@
+([Ringler et al. 2010](https://www.sciencedirect.com/science/article/pii/S0021999109006780) eqn 24)
+
+$$
+u_e^\perp = \frac{1}{d_e} \sum_{e'\in ECP(e)} w_{e,e'} \, l_e \, u_{e'}.
```
The index on `l` should be `e'`. There is also another issue. While this agrees with the TRiSK paper, it does not agree with Omega code. I believe that in MPAS-Ocean mesh files the reconstruction weights are pre-multiplied by `l_e' / d_e`. It would be good to make a note about this here. The weights also need to be added to the variables table.

mark-petersen replied:
Thanks for your detailed review! I fixed this and added text to explain what is used in the code. See the [revised section in the compiled version here](https://portal.nersc.gov/project/e3sm/mpeterse/omega/design-omega0-sw/html/design/OmegaV0ShallowWater.html#perpendicular-vector-component).

mwarusz commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,660 @@
+$$
+
+#### 3.2.8. Cell to Vertex Interpolation
+This is an area-weighted average at a vertex from the three surrounding cells. Here ${\tilde A}_v$ is the dual triangle area centered on the vertex, and ${\hat A}_{i,v}$ is the kite area of cell $i$ with the kite point at vertex $v$.
```
`A_v` is defined without the tilde in the curl equation and is missing from the variables table.

mark-petersen replied:
Done, thanks.

mwarusz reviewed (commented):
@mark-petersen I found more minor things.

vanroekel commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,679 @@
+### 2.2  Numerical method will be the TRiSK formulation
+
+The horizontal discretization will be taken from
+[Thuburn et al. 2009](https://www.sciencedirect.com/science/article/pii/S0021999109004434) and [Ringler et al. 2010](https://www.sciencedirect.com/science/article/pii/S0021999109006780), as described in the algorithmic formulation in Section 3 below. This is the same base formulation as in MPAS-Ocean. In addition, we will consider small alterations from the original MPAS-Ocean horizontal discreti ...
```
suggest a minor alteration
```
include them as options if they are of minimal additional effort
```
to 
```
include them as options if they are of minimal additional effort and improve the simulated climate.
```

mark-petersen replied:
added, thanks.

mark-petersen replied:
added, thanks.

vanroekel commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,692 @@
+
+### 2.3  Omega-0 will use MPAS format unstructured-mesh domains
+
+We will continue to use the MPAS-format netcdf files for input and output, with the same mesh variable names and dimensions. This will facilitate ease of use and interoperability between MPAS-Ocean and Omega.
```
what is mean by 'MPAS-format netcdf file'?  @xylar are MPAS outputs CF compliant now?  I also assume we won't use the same variable names right?

mark-petersen replied:
This means that we continue to use the same variables in the MPAS mesh specifications, so that MPAS meshes may be used for Omega. 

We are planning to use the same variable names for mesh variables, except that we are changing from camelCase to PascalCase (first letter caps for consistency). We can still read the same files with this case change.

vanroekel replied:
Got it, thanks for the explanation.  That sounds good as you describe.

mark-petersen commented:
@mwarusz, thank you for your detailed review. I addressed all of your comments, and the [revised document is here](https://portal.nersc.gov/project/e3sm/mpeterse/omega/design-omega0-sw/html/design/OmegaV0ShallowWater.html).

vanroekel commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,692 @@
+$$
+\mathcal{F} = - C_W \frac{(\boldsymbol{u}_W - \boldsymbol{u})\left|\boldsymbol{u}_W - \boldsymbol{u}\right|}{h}
+$$
+The drag consists of simple Rayleigh drag for spin-up as well as quadratic bottom drag. The forcing is a quadratic wind forcing (see equation 1 in [Lilly et al. 2023](https://agupubs.onlinelibrary.wiley.com/doi/full/10.1029/2022MS003327)).
```
the wind forcing seems out of place here.  The form seems fine since it isn't unlike the coupled form, but I'd suggest putting a more generic form like `tau/(rho*h` approach

mark-petersen replied:
I don't understand. Dimensionally, the coefficient of a quadratic velocity (normalized by thickness) must be non-dimensional. So the C_D and C_W are non-dimensional. Also, there is no density in a shallow water equation set. 

I was just adding some forcing and dampening terms that would transfer directly to Omega-1. 

FYI, there is a [Stommel basin test case](https://gmd.copernicus.org/articles/16/1297/2023/#section8) with an exact equilibrium solution that includes wind forcing for a shallow water test case. We listed that test as a lower-priority additional test on the [performance testing design document](https://portal.nersc.gov/project/e3sm/mpeterse/omega/design-omega0-sw/html/design/OmegaV0ShallowWater.html#inertia-gravity-wave-linearized-shallow-water-no-tracers). I'm sure we could convert that forcing function to whatever form of this coefficient we choose.

vanroekel replied:
I guess all I'm saying is that in the barotropic continuous equations the stress would be written as `tau/(rho_w*h)` and what you write in here is an assumed form of tau/rho_w, my suggestion was for the full continuous equation to leave this as `tau/h` and then in the later section specify the quadratic assumption.  Not a big deal to leave as is though.

vanroekel commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,692 @@
+$$
+so that it points 90$^o$ to the left of ${\bf F}$ in a right-handed coordinate system. 
+
+In TRiSK, the native component for the prognostic velocity field is the edge-normal velocity $u_e$. The perpendicular component is also positioned at the middle of the edge, but points 90$^o$ to the left, tangential to the edge, from one vertex to another. The perpendicular velocity component is also called the tangential velocity. It is computed diagnostically as
```
is the plan to refer to this velocity as the 'perpendicular' one?  If so we may want to make a comment that in MPAS this is the tangential velocity

mark-petersen replied:
I was planning to stay with the `TangentialVelocity` name. Here perpendicular component is just descriptive, but I can clarify that.

vanroekel replied:
Ah I see.  Thanks for that clarification

vanroekel commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,692 @@
+- C_D \frac{u_e\left|u_e\right|}{[h_i]_e}.
+$$
+
+#### 3.3.8. Wind forcing
```
given this section I would suggest the generic form above and say we choose to model the wind stress as ...

vanroekel commented:
@mark-petersen this is really nice.  I think it is so great to have this all in one place.  Thanks for all the work on it

mwarusz commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,700 @@
+\left( {\bf k} \cdot \left( \nabla \times {\bf F} \right)\right)_v
+\equiv
+{\bf k} \cdot \left( \nabla \times F_e \right)
+= \frac{1}{A_v} \sum_{e\in EV(v)} t_{e,v} \, F_e \, d_e,
```
```suggestion
= \frac{1}{{\hat A}_v} \sum_{e\in EV(v)} t_{e,v} \, F_e \, d_e,
```

mwarusz commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,700 @@
+= \frac{1}{A_v} \sum_{e\in EV(v)} t_{e,v} \, F_e \, d_e,
+$$
+
+where $A_v$ is the area of the dual mesh cell $v$, i.e. the triangle surrounding vertex $v$. Similar to $n_{e,i}$, the indicator function $t_{e,v}$ tracks whether a positive $F_e$ makes a positive or negative contribution to the curl function at vertex $v$. If the vector ${\bf k} \times {\bf n}_e$ is directed toward
```
```suggestion
where ${\hat A}_v$ is the area of the dual mesh cell $v$, i.e. the triangle surrounding vertex $v$. Similar to $n_{e,i}$, the indicator function $t_{e,v}$ tracks whether a positive $F_e$ makes a positive or negative contribution to the curl function at vertex $v$. If the vector ${\bf k} \times {\bf n}_e$ is directed toward
```

mwarusz commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,700 @@
+$$
+The normalization can be found through simple geometric arguments to be similar to above,
+$$
+A_v = \sum_{e\in EV(v)} \frac{{\bar A}_{e}}{4}.
```
Shouldn't this simply be `{\hat A}_v` ? I don't see this expression in [MPAS-O code](https://github.com/E3SM-Project/E3SM/blob/264a352acdbbdc161db76cfbd3622db8829c07bf/components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F#L1153-L1160). It just uses `areaTriangle`.

mark-petersen replied:
Yes. I updated it to `{\hat A}_v`. I just meant to show here the equivalence of the two sides in area. I updated the words to make that more clear.

mwarusz commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,700 @@
+| $x_v$ | point | Location of center of dual-mesh cells (Vertex)| (xVertex, yVertex, zVertex) | (lonVertex, latVertex) |
+| $d_e$ | line | segment distance between neighboring $x_i$ locations | dcEdge | |
+| $l_e$ | line | segment distance between neighboring $x_v$ locations | dvEdge | |
+| $P_i$ | cell | A cell on the primal mesh | | |
+| $D_v$ | cell | A cell on the dual-mesh | | |
```
```suggestion
| $P_i$ | cell | A cell on the primal mesh | areaCell | |
| $D_v$ | cell | A cell on the dual-mesh | areaTriangle | |
```

mwarusz commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,700 @@
+A_v = \sum_{e\in EV(v)} \frac{{\bar A}_{e}}{4}.
+$$
+
+The computation of $K_v$ can be found in the MPAS-Ocean code with the `kineticEnergyVertexOnCells` variable. The final kinetic energy at the cell center may be chosen to be either $K_i$ in the first equation above, or a linear combination of the two. In MPAS-Ocean, the flag `config_include_KE_vertex` uses the value
```
```suggestion
The computation of $K_v$ can be found in the MPAS-Ocean code with the `kineticEnergyVertex` variable. The final kinetic energy at the cell center may be chosen to be either $K_i$ in the first equation above, or a linear combination of `K_i` and `K_v` interpolated to cell centers. In MPAS-Ocean, the flag `config_include_KE_vertex` uses the value
```

mwarusz commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,700 @@
+The computation of $K_v$ can be found in the MPAS-Ocean code with the `kineticEnergyVertexOnCells` variable. The final kinetic energy at the cell center may be chosen to be either $K_i$ in the first equation above, or a linear combination of the two. In MPAS-Ocean, the flag `config_include_KE_vertex` uses the value
+
+$$
+\frac{5}{8} K_i + \frac{3}{8} [K_v]_i
```
Vertex to cell interpolation (`[K_v]_i`) is not defined in the document.

mark-petersen replied:
Added that. thanks.

mwarusz reviewed (commented):
@mark-petersen Thanks for addressing all of my previous comments. This looks almost ready, but I do have some new suggestions.

mark-petersen commented:
@mwarusz thanks for your attention to detail, and sorry for the delay. The revised documentation is [available here](https://portal.nersc.gov/project/e3sm/mpeterse/omega/design-omega0-sw/html/design/OmegaV0ShallowWater.html). I added a new figure to show exactly what each area reagion is to help future readers, below the [variable definitions](https://portal.nersc.gov/project/e3sm/mpeterse/omega/design-omega0-sw/html/design/OmegaV0ShallowWater.html#variable-definitions).

mwarusz commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,711 @@
+[Ringler et al. 2010](https://www.sciencedirect.com/science/article/pii/S0021999109006780) equation 63,
+
+$$
+K_i = \frac{1}{A_i} \sum_{e\in EC(c)} \frac{{\bar A}_{e}}{4}u_e^2.
```
```suggestion
K_i = \frac{1}{A_i} \sum_{e\in EC(i)} \frac{{\bar A}_{e}}{4}u_e^2.
```

mwarusz commented on components/omega/doc/design/OmegaV0ShallowWater.md:
```diff
@@ -0,0 +1,711 @@
+
+Here ${\bar A}_{e}$ is the rectangle about the edge defined by ${\bar A}_{e} = l_e d_e$ so that
+$$
+A_i = \sum_{e\in EC(c)} \frac{{\bar A}_{e}}{4}
```
```suggestion
A_i = \sum_{e\in EC(i)} \frac{{\bar A}_{e}}{4}
```

mwarusz reviewed (approved):
Thanks @mark-petersen for your hard work on this document. The added figure is very helpful. The only issue I found are two typos where `c` is used as cell index instead of `i`, so I think I can safely approve.

mark-petersen commented:
rebased to handle conflicts.
